### PR TITLE
Testing context model

### PR DESCRIPTION
**Pull request by omangin**
Branch and PR for @alancxchen's work on the repository.

### CODE REVIEW

File: hrc_speech_prediction/tests/test_context_model.py [hunk 1]
```diff
@@ -7,20 +7,60 @@
...
+        self.context = ['a','b','c']
+        self.contexts = [['a','b','c'],['d','e','f'],['g','h','i'],['a','g','e']]
         pass
 
     def test_empty_has_zero_n_children(self):
-        raise NotImplementedError
+        self.x = Node()
+        self.assertEqual(self.x.n_children, 0)
+
+        #delete this
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Avoid committing code that is not meant to stay.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+from hrc_speech_prediction.context_model import ContextTreeModel, Node
+
+
+class TestNode(TestCase):
+
+    def setUp(self):
+        self.empty_context = []
+        self.context = ['a','b','c']
+        self.contexts = [['a','b','c'],['d','e','f'],['g','h','i'],['a','g','e']]
+        pass
```
[the remark this answers is not in the document]
Comment (omangin) [reply hrc_speech_prediction/tests/test_context_model.py]: Indeed!

File: hrc_speech_prediction/tests/test_context_model.py [hunk 3]
```diff
@@ -7,20 +7,60 @@
...
-        raise NotImplementedError
+        self.x = Node()
+        self.assertEqual(self.x.n_children, 0)
+
+        #delete this
+        self.x.get_or_add_node([])
+        print("Hello")
+        print(self.x.n_children)
+        #no deletion code
+        # self.x.get_or_add_node(self.contxt)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Same thing here: only push the code to the branch that is meant to work.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 4]
```diff
@@ -7,20 +7,60 @@
...
 
     def setUp(self):
+        self.empty_context = []
+        self.context = ['a','b','c']
+        self.contexts = [['a','b','c'],['d','e','f'],['g','h','i'],['a','g','e']]
         pass
 
     def test_empty_has_zero_n_children(self):
-        raise NotImplementedError
+        self.x = Node()
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Here you can use a local variable instead of creating an attribute of the TestCase since the node will only be used locally.
```python
x = Node()
self.assertEqual(x.n_children, 0)
```

File: hrc_speech_prediction/tests/test_context_model.py [hunk 5]
```diff
@@ -7,20 +7,60 @@
...
+        print("Hello")
+        print(self.x.n_children)
+        #no deletion code
+        # self.x.get_or_add_node(self.contxt)
 
     def test_nonempty_has_n_children(self):
-        raise NotImplementedError
+        self.x = Node()
+        self.x.get_or_add_node(self.empty_context)
+        self.assertEqual(self.x.n_children, 0)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: This is not the condition that the test is supposed to test.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 6]
```diff
@@ -7,20 +7,60 @@
...
+        #delete this
+        self.x.get_or_add_node([])
+        print("Hello")
+        print(self.x.n_children)
+        #no deletion code
+        # self.x.get_or_add_node(self.contxt)
 
     def test_nonempty_has_n_children(self):
-        raise NotImplementedError
+        self.x = Node()
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Same comment for the `self.x`. Here you typically want to use a local variable. If it makes sense to initialize a Node always the same way for all tests, you can do that but then it should be done in the `setUp` method.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 7]
```diff
@@ -7,20 +7,60 @@
...
+
 
-    def test_get_or_add_node_increments_count(self):
-        raise NotImplementedError
 
+    def test_add_branch_increments_count(self):
+        self.x = Node()
+        self.assertEqual(self.x._count,0)
+        self.branch = self.x.add_branch(self.context)
+        self.assertEqual(self.      branch._count, 1)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Space?

File: hrc_speech_prediction/tests/test_context_model.py [hunk 8]
```diff
@@ -7,20 +7,60 @@
...
+        self.x = Node()
+        self.x.get_or_add_node(self.context)
+
+        self.assertTrue(self.context[0] in self.x._children.keys())
+        self.assertTrue(self.context[1] in self.x._children[self.context[0]]._children.keys())
+        self.assertTrue(self.context[2] in self.x._children[self.context[0]]._children[self.context[1]]._children.keys())
+
+
 
-    def test_get_or_add_node_increments_count(self):
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Why did you remove this one?

File: hrc_speech_prediction/tests/test_context_model.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    def test_n_children(self):
+        self.assertEqual(self.model.root.n_children, 2)
+
+    def test_fit_raises_exception_on_context_action_mismatch(self):
+    #     #test that this raises a value error
+        self.assertRaises(ValueError, self.model.fit, [[], ['b'], ['c']], ['a','b'])
+
+
+        # raise NotImplementedError
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Remove commented code and space.

**Comment by omangin** [comment]
@alancxchen Good start! 👍 Please see my comments on the code.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    def test_empty_has_zero_n_children(self):
+        x = Node()
+        self.assertEqual(x.n_children, 0)
+
+    def test_nonempty_has_n_children(self):
+        x = Node()
+        x.get_or_add_node(self.empty_context)
+        self.assertEqual(x.n_children, 0)
+
```
[the remark this answers is not in the document]
Comment (alecive) [reply hrc_speech_prediction/tests/test_context_model.py]: @alancxchen the trailing whitespace issue can be easily solved if you enable the proper setting in your text editor / IDE!!

File: hrc_speech_prediction/tests/test_context_model.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        self.assertEqual(branch._count, 1)
+        branch = x.add_branch(self.context)
+        self.assertEqual(branch._count, 2)
+
+        self.assertEqual(x._children['a']._children['b']._count, 0)
+        branch = x.add_branch(['a','b'])
+        self.assertEqual(branch._count, 1)
+
+class TestContextTreeModel(TestCase):
+    def setUp(self):
```
Comment (alecive) [inline hrc_speech_prediction/tests/test_context_model.py:49]: I know that this was the case also in the previous version, but for uniformity I would change the name of this function from `setUp` to `set_up`

File: hrc_speech_prediction/tests/test_context_model.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
+from unittest import TestCase
+
+from hrc_speech_prediction.data import ALL_ACTIONS
+from hrc_speech_prediction.context_model import ContextTreeModel, Node
+
+
+class TestNode(TestCase):
+
+    def setUp(self):
```
Comment (alecive) [inline hrc_speech_prediction/tests/test_context_model.py]: I know that this was the case also in the previous version, but for uniformity I would change the name of this function from `setUp` to `set_up`

**Review by alecive — DISMISSED** [review]

File: hrc_speech_prediction/tests/test_context_model.py (hunk 11 above)
Comment (omangin) [reply hrc_speech_prediction/tests/test_context_model.py:49]: That would indeed make more sense. Unfortunately, this is the name of the method that the `unittest` library is calling before running each test so we have to stick with it… And, yes, Python standard libraries are often very bad at abiding by the language's own standards!

File: hrc_speech_prediction/tests/test_context_model.py (hunk 12 above)
Comment (omangin) [reply hrc_speech_prediction/tests/test_context_model.py]: Same comment as below.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        x.get_or_add_node(['b','c'])
+        self.assertEqual(x.n_children, 2)
+
+        self.assertEqual(x.n_children, len(x._children))
+
+    def test_add_new_node_adds_child(self):
+        x = Node()
+        x.get_or_add_node(self.context)
+
+        self.assertTrue(self.context[0] in x._children.keys())
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Use the public attribute `x.seen_children` here and in the following.

It always better to test the public methods so that the tests remain the same if the implementation changes.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertTrue(self.context[1] in x._children[self.context[0]]._children.keys())
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: In addition to the previous comment, also use the public method for get `x.get_or_add_node(self.context[:1])` after asserting that the child is there (which is done by the previous line). Same in the next one.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    def test_fit_raises_exception_on_context_action_mismatch(self):
+        self.assertRaises(ValueError, self.model.fit, [[], ['b'], ['c']], ['a','b'])
+
+    def test_fit_empty_lists_does_nothing(self):
+        self.old_modelroot = self.model.root
+        self.model.fit([], [])
+        self.assertEqual(self.model.root.n_children, self.old_modelroot.n_children)
+        for i in range(len(self.model.root._children.keys())):
+            key = self.model.root._children.keys()[i]
+            self.assertEqual(self.model.root._children[key]._count, self.old_modelroot._children[key]._count)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: First of all, you can directly write your `for` loop on the keys so that it becomes:
```python
for key in self.model.root._children.keys():
    self.assertEqual(self.model.root._children[key]._count, self.old_modelroot._children[key]._count)
```
But actually iterating over a dictionary is the same as iterating over its keys so you can simply have:
```python
for key in self.model.root._children:  # or similarly: self.model.root.seen_children
    self.assertEqual(self.model.root._children[key]._count, self.old_modelroot._children[key]._count)
```
Anyway, this code is only testing one layer of the tree. Alsi, assuming the copy was really a deep copy, it would raise an error (instead of a failure) if the new tree has some new child at some point because you need to test first that the key is in the old tree children.

You could address that by manually loop over the four levels that the tree has, but this is not very nice, so instead, first write an `__equal__` method in the node class (and add tests for it) so that you can simply test `assertEqual(old_model.root, new_model.root)`.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        self.model.fit([[], [], ["a"], ["z", "b", "c"]], ["a", "a", "b", "d"])
+
+    def test_n_children(self):
+        self.assertEqual(self.model.root.n_children, 2)
+
+    def test_fit_raises_exception_on_context_action_mismatch(self):
+        self.assertRaises(ValueError, self.model.fit, [[], ['b'], ['c']], ['a','b'])
+
+    def test_fit_empty_lists_does_nothing(self):
+        self.old_modelroot = self.model.root
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: This is not actually copying the object but just creating a reference to it. Hence you are basically testing that an object is equal to itself. Here you want a **deep copy** of the model. Use [`copy.deepcopy`](https://docs.python.org/2/library/copy.html) to get it.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertEqual(self.model.root.n_children, self.old_modelroot.n_children)
+        for i in range(len(self.model.root._children.keys())):
+            key = self.model.root._children.keys()[i]
+            self.assertEqual(self.model.root._children[key]._count, self.old_modelroot._children[key]._count)
+   
+    def test_fit_one(self):
+        fit_one = ContextTreeModel(ALL_ACTIONS)
+        fit_one.fit([["a"]], ["b"])
+        self.assertEqual(fit_one.root.seen_children, ["a"])
+        self.assertTrue("b" in fit_one.root._children["a"].seen_children)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py:71]: Use `assertIn(a, b)` instead of `assertTrue(a in b)`.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        x = Node()
+        self.assertEqual(x._count,0)
+        branch = x.add_branch(self.context)
+        self.assertEqual(branch._count, 1)
+        branch = x.add_branch(self.context)
+        self.assertEqual(branch._count, 2)
+        self.assertEqual(x._children['a']._children['b']._count, 0)
+        branch = x.add_branch(['a','b'])
+        self.assertEqual(branch._count, 1)
+
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py:47]: > Surround top-level function and class definitions with two blank lines.
> Method definitions inside a class are surrounded by a single blank line.

https://www.python.org/dev/peps/pep-0008/#blank-lines

File: hrc_speech_prediction/tests/test_context_model.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        x = Node()
+        self.assertEqual(x.n_children, 0)
+
+    def test_nonempty_has_n_children(self):
+        x = Node()
+        x.get_or_add_node(self.empty_context)
+        self.assertEqual(x.n_children, 0)
+      
+        x.get_or_add_node(self.context)
+        self.assertEqual(x.n_children,1)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: The convention is to have a space after commas in lists of arguments `(a, b)`. (There are many of these to fix. You may want to use a [linter](https://www.google.com/search?q=python+linter).)

File: hrc_speech_prediction/tests/test_context_model.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+   
+    def test_fit_one(self):
+        fit_one = ContextTreeModel(ALL_ACTIONS)
+        fit_one.fit([["a"]], ["b"])
+        self.assertEqual(fit_one.root.seen_children, ["a"])
+        self.assertTrue("b" in fit_one.root._children["a"].seen_children)
+
+
+    def test_fit_several(self):
+        self.assertTrue("a" and "z" in self.model.root.seen_children)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: `"a" and "z" in blahblah` means `("a") and ("z" in blahblah)`. Hence this is not actually testing what you want. It passes because `"a"` is nonempty and evaluates as `True`.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertTrue("b" in self.model.root._children["a"].seen_children)
+        self.assertTrue("b" in self.model.root._children["z"].seen_children)
+        self.assertTrue("d" in self.model.root._children["z"]._children["b"]._children["c"].seen_children)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: `assertIn`

File: hrc_speech_prediction/tests/test_context_model.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+        
+        self.assertFalse("c" in self.model.root.seen_children)
+        self.assertEqual(2, self.model.root._children["a"]._count)
+        self.assertEqual(1, self.model.root._children["a"]._children["b"]._count)
+
+    def test_total_of_counts_incremented_by_one_after_fit_one(self):
+        fit_one = ContextTreeModel(ALL_ACTIONS)
+        old_sum = fit_one.root.sum_children_counts
+        fit_one.fit([["a"]], ["b"])
+        self.assertEqual(fit_one.root.sum_children_counts + 1, old_sum)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Are you sure this is what you want to test? (Hint: the code does not give the correct result anyway.)

**Review by omangin — COMMENTED** [review]
@alancxchen It is close to be there but a few of the test need fixes (some requiring implementing some new methods).

Please also check you white spaces and blank lines as mentioned.

**Comment by omangin** [comment]
@alancxchen I have just merged automatic testing in the branch. As you already know, it is not passing right now but should before merging the branch.

**Comment by alecive** [comment]
@alancxchen you have to fix the compilation errors on travis, as well as the codacy comments.

**Comment by alecive** [comment]
@omangin can we disable the DEBUG mode in matplotlib when running travis? It prints a lot of useless things that make inspection of the logs difficult.

File: hrc_speech_prediction/hrc_speech_prediction/context_model.py [hunk 23]
```diff
@@ -55,6 +55,27 @@ def _increment_count(self):
         self._count += 1.0
         return self
 
+    def __eq__(self, other):
+        if self.n_children == 0 and other.n_children == 0:
+            return True
```
Comment (omangin) [inline hrc_speech_prediction/hrc_speech_prediction/context_model.py]: Here we want the equality to mean that both the children and the number of counts are the same. The current code does not address the latter for the root.

File: hrc_speech_prediction/hrc_speech_prediction/context_model.py [hunk 24]
```diff
@@ -55,6 +55,27 @@ def _increment_count(self):
...
+                # test for same count
+                if self._children[key]._count != other._children[key]._count:
+                    return False
+                # test for same children, can't be == cuz not defined yet
+                if not self._children[key].__eq__(other._children[key]):
+                    return False
+
+        else:
+            return False
+        return True
```
Comment (omangin) [inline hrc_speech_prediction/hrc_speech_prediction/context_model.py]: Here `return self._children == other._children` would do the job since the equality on the dictionaries will itself test the equality of the children node.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 25]
```diff
@@ -101,3 +98,16 @@ def test_empty_predicts_uniform(self):
...
+    def test_equals(self):
+        empty = Node()
+        empty2 = Node()
+        self.assertEqual(empty, empty2)
+        newTree = ContextTreeModel(ALL_ACTIONS)
+        self.assertFalse(newTree.root == self.model.root)
+        newTree.fit([[], [], ["a"]], ["a", "a", "b"])
+        self.assertFalse(newTree.root == self.model.root)
+        newTree.fit([ ["z", "b", "c"]], ["d"])
+        self.assertTrue(newTree.root ==self.model.root)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Add a test that they are not equal when the children are the same but not the counts.

Also fix the spacing.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 26]
```diff
@@ -101,3 +98,16 @@ def test_empty_predicts_uniform(self):
...
+
+
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: Avoid multiple blank lines at the end of a file.

File: hrc_speech_prediction/tests/test_context_model.py [hunk 27]
```diff
@@ -101,3 +98,16 @@ def test_empty_predicts_uniform(self):
         prediction = empty.predict(["A"])
         for i in range(len(prediction)-1):
             self.assertEqual(prediction[i], prediction[i+1])
+
+    def test_equals(self):
+        empty = Node()
+        empty2 = Node()
+        self.assertEqual(empty, empty2)
+        newTree = ContextTreeModel(ALL_ACTIONS)
+        self.assertFalse(newTree.root == self.model.root)
```
Comment (omangin) [inline hrc_speech_prediction/tests/test_context_model.py]: `assertNotEqual`

**Review by omangin — COMMENTED** [review]
Almost there! Mostly minor comments to be fixed.

You however didn't address this comment https://github.com/ScazLab/hrc_speech_prediction/pull/2#discussion_r177241928 (you are testing `new + 1 == old` which should not pass). Please do.

As mentioned by @alecive, make sure the Travis build passes after that.
